### DISCUSSION TITLE
LlamaIndex integration (Python)

### DISCUSSION THREAD

**Discussion by marcklingen** [langfuse/langfuse#828] [Ideas]
Many users approached @maxdeichmann, @clemra and me regarding a Llamaindex integration. Keeping this request here to track the overall interest and discuss a potential implementation.

From #819 
> I would like to use Langfuse with [Llamaindex](https://github.com/run-llama/llama_index) to track my LLM interactions. I would like to use a similar callback handler to the langchain one.
> A callback handler adjusted for the Llamaindex interface would be nice. As the library builds onto langchain, the interface is not super different.
> [Code for the interface of a Lllamaindex callback handler](https://github.com/run-llama/llama_index/blob/main/llama_index/callbacks/base_handler.py)

We are considering implementing a dedicated observation type for context embedding and retrieval to offer better RAG analytics. The Llamaindex integration would be a good occasion to implement and test these new types as well. However, it is not blocked by it as we can start with spans and move to the new types at a later time.

**Comment by brarrow** [comment]
I'm really looking forward to it.

**Comment by zzzming** [comment]
That's an awesome idea. Do you plan to add to the \callbacks folder?

**Comment by marcklingen** [comment]
We got started on the Llamaindex integration (Python for now) and the first version looks promising. We'll test it with some users/customers over the next days, email me if you are interested to try it and give feedback: marc at langfuse.com
cc @zzzming @brarrow

**Comment by marcklingen** [comment]
We've just released the first alpha version of this integration (mostly stable but some things might still change before we add it to the documentation and the llamaindex package -> not covered by semver yet, interfaces might still change slightly).

Here are previews of the docs in case you are interested to try it:
- [Changelog post](https://langfuse-docs-git-feat-adds-llamaindex-integration-langfuse.vercel.app/changelog/2024-02-20-llama-index-integration) (preview)
- [Docs](https://langfuse-docs-git-feat-adds-llamaindex-integration-langfuse.vercel.app/docs/integrations/llama-index/overview) (preview) 
- [Jupyter notebook](https://github.com/langfuse/langfuse-docs/blob/feat-adds-llamaindex-integration/cookbook/integration_llama-index.ipynb) with an end to end example and links to public previews

Any feedback? Please let me know here or email me: marc at langfuse com

**Comment by spreeni** [comment]
Hey @marcklingen, those are great news first of all! I just tested around a bit, needed a bit to include the huge refactoring from llamaindex 0.9 to 0.10.

Your demo script works for me, but within my RAG application, it did not work initially. After testing a bit, I realized I needed to replace all `ServiceContext` usages (now deprecated), then it propagates basic info. 

If I set the callback handler individually for each created LlamaIndex instance, I can see retrieval results for the `VectorStoreIndex`, and generation outputs. For queries this already works completely, for chats it does not display the answer correctly for me.

```python
Settings.embed_model = OpenAIEmbedding(**embedding_args, callback_manager=Settings.callback_manager)
Settings.llm = OpenAI(..., callback_manager=Settings.callback_manager)
index = VectorStoreIndex.from_vector_store(..., callback_manager=Settings.callback_manager)
ContextChatEngine.from_defaults(
    retriever=index.as_retriever(),
    callback_manager=Settings.callback_manager,
)
```
<img width="1416" alt="Screenshot 2024-02-22 at 12 32 45" src="https://github.com/langfuse/langfuse/assets/35889034/f16df1f5-dfdb-41a7-baf7-adf9950948e1">

**Reply by marcklingen** [reply]
We're currently working on an improvement. Do you stream chat responses?

**Reply by spreeni** [reply]
No, this is without streaming responses. Tbh, some of these problems are probably linked to the recent massive refactoring on the side of llama-index and it not always adapting the settings from `core.Settings`.

**Reply by hassiebp** [reply]
Hey @spreeni - does the issue persist even if you follow the [latest implementation guide here](https://docs.llamaindex.ai/en/stable/examples/chat_engine/chat_engine_context.html#chat-engine-context-mode)? 🤔  
```python 
from llama_index.core.callbacks import CallbackManager
from langfuse.llama_index import LlamaIndexCallbackHandler
 
langfuse_callback_handler = LlamaIndexCallbackHandler()
Settings.callback_manager = CallbackManager([langfuse_callback_handler])

index = VectorStoreIndex.from_vector_store(...)
chat_engine = index.as_chat_engine()
chat_engine.chat("Whassup?")
```

**Reply by jimgoo** [reply]
@marcklingen Are streaming chat responses supported yet for this integration? With langfuse-python 2.20.0, using the new way of setting it up mentioned by @hassiebp above, I'm getting this error when streaming:

```
ERROR:langfuse:Exception in adding task cannot pickle 'generator' object
Traceback (most recent call last):
  File "/Users/jimgoo/miniconda3/envs/langfuse-test/lib/python3.11/site-packages/langfuse/task_manager.py", line 235, in add_task
    json.dumps(event, cls=EventSerializer)
```

The generator that's causing the problem is for the streaming response here in the event:

```
{'id': 'abf522a8-d108-4414-b961-dd6cfb40b8e3', 'type': 'trace-create', 'body': {'id': '1f49ec02-5eb3-4797-8cc5-ae651ea19257', 'input': 'Hello how are you?', 'output': {<EventPayload.RESPONSE: 'response'>: StreamingResponse(response_gen=<generator object stream_chat_response_to_tokens.<locals>.gen at 0x2b5c27370>
```

I'd submit a PR if I could figure it out, open to suggestion on how to go about fixing.

**Reply by hassiebp** [reply]
Thanks for your report @jimgoo! Could you please share a code snippet of the LlamaIndex call that triggered the erroring trace?🙏🏾

**Reply by hassiebp** [reply]
@jimgoo Assuming you've been using `index.as_query_engine(streaming=True)`, I have prepared a [fix in this PR](https://github.com/langfuse/langfuse-python/pull/476). Will ping you once merged & released :)

**Reply by jimgoo** [reply]
That’s correct I’m using that exact thing. When I get back to laptop I’ll post the exact code snippet here and try out your PR. Thanks!

**Reply by jimgoo** [reply]
I posted my full code on the linked PR. After trying that branch, it fixed the picking error, but I wasn’t able to iterate over the streaming response generator to print it, maybe b/c the string conversion consumes all the events.

**Reply by hassiebp** [reply]
Thanks - [fixed here](https://github.com/langfuse/langfuse-python/pull/476#issuecomment-1987284722) 🙂

**Reply by jimgoo** [reply]
Thanks that worked!

**Comment by marcklingen** [comment]
Thanks everyone for testing this and providing feedback!

We've iterated quite a bit on the implementation to support all Langfuse platform features conveniently. Have a look here: https://langfuse.com/changelog/2024-02-27-llama-index-integration

If you have any further feedback, please add it here.

Please create issues if you experience any bugs when using the integration so we can fix them asap.

**Comment by marcklingen** [comment]
Create a new idea here to track interest for a **LlamaIndex JS/TS** integration: https://github.com/orgs/langfuse/discussions/1291
